core/shared/src/main/scala/scalaz/zio/STM.scala

```diff
@@ -0,0 +1,226 @@
+          case (journal, Left(error)) => (journal, Left(error))
+          case (journal, Right(a))    => f(a).run(journal)
+      }
+    )
+
+  /**
+   * Tries this computation first, and if it fails, tries the other computation.
+   */
+  def orElse[E2, A1 >: A](that: => STM[E2, A1]): STM[E2, A1] =
+    self.attempt.flatMap(either => either.fold(_ => that, STM.point(_)))
+
+  /**
+   * Surfaces any errors at the value level, where they can be recovered from.
+   */
+  def attempt: STM[Nothing, Either[E, A]] =
+    STM(
+      journal =>
+        self.run(journal) match {
+          case (journal, value) => (journal, Right(value))
+      }
+    )
+
+  /**
+   * Sequentially zips this value with the specified one.
+   */
+  def seq[E1 >: E, B](that: => STM[E1, B]): STM[E1, (A, B)] =
+    self.flatMap(a => that.map(b => (a, b)))
+}
+
+object STM {
+
+  object internal {
+    def freshIdentity(): Long = counter.incrementAndGet()
+
+    private val counter: AtomicLong = new AtomicLong()
+
+    val semaphore = new java.util.concurrent.Semaphore(1)
+
+    abstract class Entry {
+      type A
+      val tVar: TVar[A]
+      val newValue: A
+      val expectedVersion: Long
+
+      def isValid: Boolean = tVar.version == expectedVersion
+      def commit(): Unit   = tVar.value = newValue
+    }
+
+    object Entry {
+
+      def apply[A0](tVar0: TVar[A0], newValue0: A0, expectedVersion0: Long): Entry = new Entry {
+        type A = A0
+        val tVar            = tVar0
+        val newValue        = newValue0
+        val expectedVersion = expectedVersion0
+      }
+    }
+  }
+
+  import internal._
+
+  /**
+   * A variable that can be modified as part of a transactional computation.
+   */
+  class TVar[A] private (identity: Long, @volatile var version: Long, @volatile var value: A) {
+    self =>
+
+    final def get: STM[Nothing, A] =
+      STM(journal => {
+        val entry = journal.getOrElse(identity, Entry(self, value, version))
+
+        (journal.updated(identity, entry), Right(value))
+      })
+
+    final def set(newValue: A): STM[Nothing, Unit] =
+      STM(journal => {
+        val entry = journal.getOrElse(identity, Entry(self, newValue, version))
+
+        (journal.updated(identity, entry), Right(()))
+      })
+
+    final def update(f: A => A): STM[Nothing, Unit] =
+      get.flatMap(a => set(f(a)))
+  }
+
+  object TVar {
+    final def apply[A](initialValue: => A): STM[Nothing, TVar[A]] =
+      STM(journal => {
+        val identity = freshIdentity()
+
+        val tVar = new TVar(identity, 0, initialValue)
+
+        (journal.updated(identity, Entry(tVar, initialValue, 0)), Right(tVar))
+      })
+  }
+
+  final def point[A](a: => A): STM[Nothing, A] =
+    STM(journal => (journal, Right(a)))
+
+  final def atomically[E, A](stm: STM[E, A]): IO[E, A] =
+    IO.absolve(IO.sync[Either[E, A]] {
+      import internal.semaphore
+
+      var retry = true
+
+      val (journal, value) = stm.run(Map.empty)
+
+      while (retry) {
```

Comment:
Can this not be done with the retry functionality and `bracket` on `IO`?

Reply:
@NeQuissimus This could be done with IO. I remember during design session john  mentioned since atomically has be optimized and fast he preferred while. Lets wait for @jdegoes comment as well.